### PR TITLE
feat(pnpm): don't fail 'recursive run' if no selected packages have script

### LINKED ISSUE #1763: Feature request: add ability to let `pnpm recursive run` succeed even when no scripts match
Thanks for such a great project!

This issue is just a feature request. If you think it's worth implementing, I'd be happy to take a stab at implementing it.

# Feature

Add the ability to let `pnpm recursive run foo` succeed when no packages contain a script

# Use Case

I'm using PNPM in a monorepo, and I'm using a custom script based on `git diff` that creates filters for PNPM based on packages that have changed. Basically a homegrown version of [Lerna's `--since` flag](https://github.com/lerna/lerna/tree/master/core/filter-options#--since-ref). It looks something like:

```sh
pnpm m run build $(since master)

# expands to e.g.
pnpm m run build --filter foo --filter bar
```

With pnpm's current behavior, if none of the packages that have changed include a build script, the command will fail.

In daily use, it's not a big deal, but in CI, it causes the build to fail (somewhat) unexpectedly if not all packages contain a certain CI script. In my case, the build often fails because there's a mix of browser packages, node packages, and TS packages, and only some of them contain a `build` script.

It's currently possible to work around this using `pnpm m exec`, but it's not super pretty:

```sh
pnpm m exec \
  --filter pkgs-that-might-have-script \
  -- bash -c 'if pnpm run | grep "^  build"; then pnpm run build; else true; fi'
```

Ideally, I'd like to be able to do something like:

```sh
pnpm m run build \
  --filter pkgs-that-might-have-script \
  --no-fail-on-missing-script
```

### PR DESCRIPTION
**Pull request by bdchauvette**
This changes the behavior of of `pnpm recursive run ...` so that it no
longer throws an error when run against a subset of packages and none of
the selected packages have the provided script.

Previously, running a command like `pnpm m run --filter foo build` would
throw an error if the `foo` package did not have a `build` script, even
if other workspace packages have a `build` script.

With this change, running `pnpm recursive run --filter foo build` will
not throw an error.

Note that if the provided package selectors match all packages (e.g.
`'*'`), then `pnpm recursive run` will fail if none of the packages have
the script. This is because the behavior depends on whether we're
running against a subset of packages or not, _not_ whether any package
selectors were provided.

Closes #1763

### CODE REVIEW

**Comment by bdchauvette** [comment]
Not sure why the Node 10 build failed on travis. Looks like it's failing on this:

https://github.com/pnpm/pnpm/blob/8e83de66f4f12dc04d84c1178870e6ad42377ced/packages/pnpm/test/install/misc.ts#L208

https://travis-ci.org/pnpm/pnpm/jobs/598443394#L7490

:man_shrugging:

File: packages/pnpm/test/recursive/run.ts [hunk 1]
```diff
@@ -157,6 +157,116 @@ test('`pnpm recursive run` fails if none of the packaegs has the desired command
...
+    {
+      name: 'project-3',
+      version: '1.0.0',
+
+      dependencies: {
+        'json-append': '1',
+        'project-1': '1'
+      },
+      scripts: {
+        build: `node -e "process.stdout.write('project-3')" | json-append ../output.json`,
```
Comment (zkochan) [inline packages/pnpm/test/recursive/run.ts]: these builds scripts are not used in this test, so remove them and also remove the json-append deps

File: packages/pnpm/test/recursive/run.ts [hunk 2]
```diff
@@ -157,6 +157,116 @@ test('`pnpm recursive run` fails if none of the packaegs has the desired command
...
```
Comment (zkochan) [inline packages/pnpm/test/recursive/run.ts]: remove the scripts and json-append

File: packages/pnpm/src/cmd/recursive/run.ts [hunk 3]
```diff
@@ -84,7 +85,7 @@ export default async <T> (
     )))
   }
 
-  if (scriptName !== 'test' && !hasCommand) {
+  if (scriptName !== 'test' && !hasCommand && opts.allPackagesAreSelected) {
```
Comment (zkochan) [inline packages/pnpm/src/cmd/recursive/run.ts]: Do we print some info message in this case? I think something should be printed with logger.info

Like `None of the selected packages has a "${scriptName}" script`

**Comment by zkochan** [comment]
> Not sure why the Node 10 build failed on travis. Looks like it's failing on this:

That's just an unstable test. I reran it

File: packages/pnpm/src/cmd/recursive/run.ts (hunk 3 above)
Comment (bdchauvette) [reply packages/pnpm/src/cmd/recursive/run.ts]: What should the `prefix` be for the log? I see that `workspacePrefix` is used in the default export for `recursive.ts`:

https://github.com/pnpm/pnpm/blob/d2dc3f29f37141bdd746cb349f7ffedea4849165/packages/pnpm/src/cmd/recursive/index.ts#L80

But the normalization on line 76 isn't put back into `opts.workspacePrefix`, so it's still potentially `undefined` when passed into the run command:

https://github.com/pnpm/pnpm/blob/d2dc3f29f37141bdd746cb349f7ffedea4849165/packages/pnpm/src/cmd/recursive/index.ts#L76

If we want to use `opts.workspacePrefix`, I think I could do any of the following:

- store the normalized prefix on `opts`, so its definitely defined (my preference)
- redo the normalization in `run.ts`: `{ prefix: opts.workspacePrefix || process.cwd() }`
- fall back to `opts.prefix`: `{ prefix: opts.workspacePrefix || opts.prefix }`

Comment (zkochan) [reply packages/pnpm/src/cmd/recursive/run.ts]: > store the normalized prefix on opts, so its definitely defined (my preference)

that sounds right

File: packages/pnpm/test/recursive/run.ts [hunk 4]
```diff
@@ -150,13 +175,53 @@ test('`pnpm recursive run` fails if none of the packaegs has the desired command
...
+
+      dependencies: {
+        'project-1': '1'
+      },
+    },
+    {
+      name: 'project-0',
+      version: '1.0.0',
+
+      dependencies: {},
```
Comment (zkochan) [inline packages/pnpm/test/recursive/run.ts]: You can remove this

Comment (bdchauvette) [reply packages/pnpm/test/recursive/run.ts]: Removed in b047b41

**Comment by zkochan** [comment]
I added you to collaborators as you did a successful pull request with tests and everything

**Comment by bdchauvette** [comment]
oh cool, thanks!

**Pull request merged**